File: Demos/Datasnap/Server/proxy/objectivec_ios81/Proxy/TDBXReader.m
//*******************************************************
//
//               Delphi DataSnap Framework
//
// Copyright(c) 1995-2011 Embarcadero Technologies, Inc.
//
//*******************************************************

#import "TDBXReader.h"
#import "DBXJsonTools.h"


@implementation TDBXReader

@synthesize columns;
@synthesize internalDataStore;


-(id) initWithParams: (TParams * )params andJSONObject: (TJSONObject *)json{
	self = [self init];
	if (self) {
		self.columns = params;
		self.internalDataStore = json;
	}
	return self;
}

-(id) init{
	self = [super init];
	if (self) {
		currentPosition = -1;
	}
	return self;	
}
-(DBXWritableValue *) getValueByIndex: (int) index{
	return [[columns getParamByIndex:index]getValue ];
}
-(DBXWritableValue *) getValueByName: (NSString *) name{
	return [[columns getParamByName:name] getValue];

}
-(bool) next{
	currentPosition++;
    TParams *tempcolumns = nil;
	@try {
		return [TParams loadParametersValues:&tempcolumns withJSON:internalDataStore andOffSet:currentPosition];
	} @catch (NSException* ex) {
			@throw ex;
			}
    columns = tempcolumns;
}
-(TJSONObject *) asJSONObject {
	return [DBXJsonTools  DBXReaderToJSONObject:self];
}
-(DBXWritableValue*) getValue:(int) position {
	return [[[self columns] getParamByIndex:position] getValue];
}
-(void) reset{
	currentPosition = -1;		
}
@end


@implementation TDBXReader (TDBXReaderCreation)
+(id) DBXReaderWithJSON:(TJSONObject *) value {
	TParams * params = [TParams createParametersWithMetadata:[value getJSONArrayForKey:@"table" ]]  ;
	
	return [ TDBXReader DBXReaderWithParams:params andJSONObject: value];
}
+(id) DBXReaderWithParams: (TParams * )params andJSONObject: (TJSONObject *)json{
	return [[TDBXReader alloc]initWithParams:params andJSONObject:json];
}


@end

File: Demos/Datasnap/Server/proxy/java_android/com/embarcadero/javaandroid/TDBXReader.java
//*******************************************************
//
//               Delphi DataSnap Framework
//
// Copyright(c) 1995-2011 Embarcadero Technologies, Inc.
//
//*******************************************************

package com.embarcadero.javaandroid;

/**
 * 
 * TDBXReader provides a unidirectional reader for a collection of database
 * rows.
 * 
 */

public class TDBXReader implements JSONSerializable, TableType {
	protected long currentPosition = -1;
	private TParams columns;
	private TJSONObject internalDataStore;

	protected void setParameters(TParams params) {
		columns = params;
	}

	public TDBXReader(TParams params, TJSONObject value) {
		super();
		internalDataStore = value;
		setParameters(params);
	}

	public TParams getColumns() {
		return columns;
	}

	public DBXWritableValue getValue(int position) {
		return columns.getParameter(position).getValue();
	}

	public DBXWritableValue getValue(String name) throws DBXException {
		return columns.getParamByName(name).getValue();
	}

	public boolean next() {
		currentPosition++;
		try {
			return TParams.LoadParametersValues(this.columns,
					this.internalDataStore, (int) currentPosition);
		} catch (Exception ex) {
			return false;
		}
	}

	public static TDBXReader createFrom(TJSONObject value) throws DBXException {
		TParams params = TParams.CreateParametersFromMetadata(value
				.getJSONArray("table"));
		TDBXReader rdr = new TDBXReader(params, value);
		return rdr;
	}

	public TJSONObject asJSONObject() throws DBXException {
		TJSONObject result = null;
		long lastPosition = currentPosition;
		try {
			reset();
			result = DBXJSONTools.DBXReaderToJSONObject(this);
		} finally {
			currentPosition = lastPosition;
		}
		return result;
	}

	public void reset() {
		currentPosition = -1;
	}

}
